module/zfs/zstd.c

```diff
@@ -0,0 +1,617 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2016-2018 by Klara Systems Inc.
+ * Copyright (c) 2016-2018 Allan Jude <allanjude@freebsd.org>.
+ */
+
+#include <sys/param.h>
+#include <sys/sysmacros.h>
+#include <sys/zfs_context.h>
+#include <sys/zio_compress.h>
+#include <sys/spa.h>
+
+#define	ZSTD_STATIC_LINKING_ONLY
+#include <zstd.h>
+#include <zstd_errors.h>
+
+#define	ZSTD_KMEM_MAGIC		0x20160831
+
+/* for BSD compat */
+#define	roundup2(x, m)	((((x) - 1) | ((m) - 1)) + 1)
+#define	__unused			__attribute__((unused))
+typedef int		 cmp_t(const void *, const void *);
+static inline char	*med3(char *, char *, char *, cmp_t *, void *);
+#ifndef MIN
+#define	MIN(a, b)	((a) < (b) ? a : b)
+#endif
+static inline void
+swapfunc(char *a, char *b, size_t es)
+{
+	char t;
+
+	do {
+		t = *a;
+		*a++ = *b;
+		*b++ = t;
+	} while (--es > 0);
+}
+
+#define	vecswap(a, b, n)				\
+	if ((n) > 0) swapfunc(a, b, n)
+#define	CMP(t, x, y) (cmp((x), (y)))
+
+static inline char *
+med3(char *a, char *b, char *c, cmp_t *cmp, void *thunk __unused)
+{
+	return CMP(thunk, a, b) < 0 ?
+	    (CMP(thunk, b, c) < 0 ? b : (CMP(thunk, a, c) < 0 ? c : a))
+	    :(CMP(thunk, b, c) > 0 ? b : (CMP(thunk, a, c) < 0 ? a : c));
+}
+
+#define	thunk NULL
+void
+zstd_qsort(void *a, size_t n, size_t es, cmp_t *cmp)
```

Comment:
You should be able to drop this, we have a `qsort()` in user and kernel space.

Reply:
not in my kernel. for userspace you're correct. but the kernel has no qsort

Reply:
The kernel does not have a qsort, though it does have a heapsort which will work just as well here.  You can do something like this for a compatibility wrapper.  I wouldn't be opposed to putting this wrapper in the `spl` for convenience.

https://github.com/zfsonlinux/zfs/blob/master/module/zcommon/zprop_common.c#L44

Reply:
see latest revision. all your comments are already resolved in the way you want it. kernel "sort" is now used for kernelspace and for userspace qsort is used